### PR TITLE
Adding support for ValueScan node in ORCA

### PR DESCRIPTION
**Pull request by khannaekta**
Postgres and thus (GPDB Planner) supports Values via an operator called ValueScan.
```
explain SELECT * FROM (VALUES (1, 'one'), (2, 'two'), (3, 'three')) AS t (num,letter);
                          QUERY PLAN
--------------------------------------------------------------
 Values Scan on "*VALUES*"  (cost=0.00..0.04 rows=1 width=36)
 Optimizer status: legacy query optimizer
(2 rows)
```
However, inside Orca we expand each row in the Values list into a Result node that projects two constants.
Thus the above query with three rows having 2 columns is represented as a plan by GPORCA as an append with three Result nodes.
Each of the Result nodes is a CTG with project elements.
```
explain SELECT * FROM (VALUES (1, 'one'), (2, 'two'), (3, 'three')) AS t (num,letter);
                   QUERY PLAN
-------------------------------------------------
 Append  (cost=0.00..0.00 rows=1 width=12)
   ->  Result  (cost=0.00..0.00 rows=1 width=12)
   ->  Result  (cost=0.00..0.00 rows=1 width=12)
   ->  Result  (cost=0.00..0.00 rows=1 width=12)
 Settings:  optimizer=on
 Optimizer status: PQO version 2.32.0
(6 rows)
```
As a first step of this effort of adding support for ValueScan node inside ORCA; this PR implements the DXL representation of the Value Scan along with the ParseHandler changes for this newly added DXL node.

### CODE REVIEW

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 1]
```diff
@@ -7530,6 +7547,53 @@ CTranslatorExprToDXL::PdxlnProjListFromConstTableGet
...
+	pcrsOutput->Include(pdrgpcrCTGOutput);
+
+	if (NULL != pdrgpcrReqOutput)
+	{
+		const ULONG ulArity = pdrgpcrReqOutput->UlLength();
+
+		for (ULONG ul = 0; ul < ulArity; ul++)
+		{
+			CColRef *pcr = (*pdrgpcrReqOutput)[ul];
+			pcrsOutput->Exclude(pcr);
```
Comment (oarap) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: you can directly exclude the list, can't you? 

```
CColRefSet::Exclude(const CColRefSet *pcrs)
```
Which was is more efficient?

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 2]
```diff
@@ -7530,6 +7547,53 @@ CTranslatorExprToDXL::PdxlnProjListFromConstTableGet
 	return pdxlnProjList;
 }
 
+// construct a project list node by creating references to the columns
+// of the given project list of the child node
+CDXLNode *
```
Comment (oarap) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: Is this code gets two sets of columsn and make them A union B and create project list from that list?

If yes, this code is unnesessarily long. 

What am I missing?

File: libnaucrates/include/naucrates/dxl/operators/CDXLScalarValuesList.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+			// conversion function
+			static
+			CDXLScalarValuesList *PdxlopConvert(CDXLOperator *pdxlop);
+
+			// does the operator return a boolean result
+			virtual
+			BOOL FBoolean
+						(
+						CMDAccessor *//pmda
```
Comment (hsyuan) [inline libnaucrates/include/naucrates/dxl/operators/CDXLScalarValuesList.h]: incorrect format

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 4]
```diff
@@ -2102,10 +2101,29 @@ CTranslatorExprToDXL::PdxlnResultFromConstTableGet
 		}
 	}
 
-	pdxlnPrL = PdxlnProjListFromConstTableGet(pdrgpcr, pdrgpcrCTGOutput, pdrgpdatum);
-	pdrgpdatum->Release();
+	// in case of multiple rows in ConstTableGet, suggests that it is a valuescan of constants.
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: reword: If CTG has multiple rows then it has to be a valuescan of constants.

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 5]
```diff
@@ -2102,10 +2101,29 @@ CTranslatorExprToDXL::PdxlnResultFromConstTableGet
...
+	// Hence a ValuesScan node is generated passing all the values as ValuesList to DXLtoPLNSTMT translator.
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: remove above sentence

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 6]
```diff
@@ -2102,10 +2101,29 @@ CTranslatorExprToDXL::PdxlnResultFromConstTableGet
...
+	// For all other cases a Result Node is created
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: else, a Result node is created

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 7]
```diff
@@ -2102,10 +2101,29 @@ CTranslatorExprToDXL::PdxlnResultFromConstTableGet
...
+	if (ulRows > 1)
+	{
+		pdxlnPrL = PdxlnProjListForValuesScan(pdrgpcr, pdrgpcrCTGOutput);
+		CDXLNode *pdxlnValuesScan = CTranslatorExprToDXLUtils::PdxlnValuesScan
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: indentation is off, can be on a single line.

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 8]
```diff
@@ -2102,10 +2101,29 @@ CTranslatorExprToDXL::PdxlnResultFromConstTableGet
...
+		pdrgpdatum->Release();
 
-	return CTranslatorExprToDXLUtils::PdxlnResult
+		return pdxlnValuesScan;
+	}
+	else
+	{
+		pdxlnPrL = PdxlnProjListFromConstTableGet(pdrgpcr, pdrgpcrCTGOutput, pdrgpdatum);
+		pdrgpdatum->Release();
+		return CTranslatorExprToDXLUtils::PdxlnResult
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: fix indentation

File: libnaucrates/src/parser/CParseHandlerValuesScan.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+	const XMLCh* const // xmlszQname
+	)
+{
+	if (0 != XMLString::compareString(CDXLTokens::XmlstrToken(EdxltokenPhysicalValuesScan), xmlszLocalname))
+	{
+		CWStringDynamic *pstr = CDXLUtils::PstrFromXMLCh(m_pphm->Pmm(), xmlszLocalname);
+		GPOS_RAISE(gpdxl::ExmaDXL, gpdxl::ExmiDXLUnexpectedTag, pstr->Wsz());
+	}
+
+	ULONG ulArity = this->UlLength();
```
Comment (vraghavan78) [inline libnaucrates/src/parser/CParseHandlerValuesScan.cpp]: const ulArity

File: libnaucrates/src/parser/CParseHandlerValuesScan.cpp [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	CParseHandlerProperties *pphProp = dynamic_cast<CParseHandlerProperties *>((*this)[0]);
+
+	// set statistics and physical properties
+	CParseHandlerUtils::SetProperties(m_pdxln, pphProp);
+
+	CParseHandlerProjList *pphPrL = dynamic_cast<CParseHandlerProjList*>((*this)[1]);
+	AddChildFromParseHandler(pphPrL);
+
+
+	for (ULONG ul = 2; ul < ulArity; ul++)
```
Comment (vraghavan78) [inline libnaucrates/src/parser/CParseHandlerValuesScan.cpp]: why 2. Can you add a comment?

File: libnaucrates/src/parser/CParseHandlerScalarValuesList.cpp [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+void
+CParseHandlerScalarValuesList::EndElement
+	(
+	const XMLCh* const, // xmlszUri,
+	const XMLCh* const, //xmlszLocalname,
+	const XMLCh* const // xmlszQname
+	)
+{
+
+	ULONG ulArity = this->UlLength();
```
Comment (vraghavan78) [inline libnaucrates/src/parser/CParseHandlerScalarValuesList.cpp]: const

File: libnaucrates/src/parser/CParseHandlerScalarValuesList.cpp [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+	for (ULONG ul=0; ul < ulArity; ul++)
```
Comment (vraghavan78) [inline libnaucrates/src/parser/CParseHandlerScalarValuesList.cpp]: ul = 0;

File: libnaucrates/src/operators/CDXLScalarValuesList.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	BOOL fValidateChildren
+	)
+	const
+{
+	const ULONG ulArity = pdxln->UlArity();
+
+	for (ULONG ul = 0; ul < ulArity; ++ul)
+	{
+		CDXLNode *pdxlnConstVal = (*pdxln)[ul];
+		GPOS_ASSERT(EdxloptypeScalar == pdxlnConstVal->Pdxlop()->Edxloperatortype());
```
Comment (vraghavan78) [inline libnaucrates/src/operators/CDXLScalarValuesList.cpp]: should we be checking if it is a scalar constant and not just a scalar?

Comment (vraghavan78) [inline libnaucrates/src/operators/CDXLScalarValuesList.cpp]: if you want to accommodate cast then target that too. Scalar is too broad. What do u think?

File: libnaucrates/src/operators/CDXLPhysicalValuesScan.cpp [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	BOOL fValidateChildren
+	)
+const
+{
+	GPOS_ASSERT(EdxloptypePhysical == pdxln->Pdxlop()->Edxloperatortype());
+
+	const ULONG ulArity = pdxln->UlArity();
+	GPOS_ASSERT(EdxlValIndexSentinel <= ulArity);
+
+	for (ULONG ul = 0; ul < ulArity; ul++)
```
Comment (vraghavan78) [inline libnaucrates/src/operators/CDXLPhysicalValuesScan.cpp]: I am confused, CDXLPhysicalValueScan has value list correct. Shouldn't that be the check?

File: libnaucrates/include/naucrates/dxl/operators/CDXLScalarValuesList.h [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+			void SerializeToDXL(CXMLSerializer *pxmlser, const CDXLNode *pdxln) const;
+
+			// conversion function
+			static
+			CDXLScalarValuesList *PdxlopConvert(CDXLOperator *pdxlop);
+
+			// does the operator return a boolean result
+			virtual
+			BOOL FBoolean
+						(
```
Comment (vraghavan78) [inline libnaucrates/include/naucrates/dxl/operators/CDXLScalarValuesList.h]: single line

File: libnaucrates/include/naucrates/dxl/operators/CDXLScalarValuesList.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+	class CDXLScalarValuesList : public CDXLScalar
+	{
+		private:
+
+			// private copy ctor
+			CDXLScalarValuesList(CDXLScalarValuesList&);
+
+		public:
+			// ctor/dtor
+			CDXLScalarValuesList
```
Comment (vraghavan78) [inline libnaucrates/include/naucrates/dxl/operators/CDXLScalarValuesList.h]: single line?

File: libnaucrates/src/operators/CDXLScalarValuesList.cpp (hunk 13 above)
Comment (oarap) [inline libnaucrates/src/operators/CDXLScalarValuesList.cpp]: There are only three operator types:
```
	enum Edxloptype
	{
		EdxloptypeLogical,
		EdxloptypePhysical,
		EdxloptypeScalar
	};
```

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp (hunk 2 above)
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: I agree, why not call CTranslatorExprToDXL::PdxlnProjList

File: libnaucrates/src/operators/CDXLPhysicalValuesScan.cpp (hunk 14 above)
Comment (bhuvnesh2703) [inline libnaucrates/src/operators/CDXLPhysicalValuesScan.cpp]: Value Scan has the below child nodes
1. Project List
2. Scalar Value List
    2.1 Const Value

Inside the for loop assert for proj list, scalar value list and their children. So we do check for value list as well. Please let us know if we are missing something.

**Pull request merged**